Recast/Source/RecastMeshDetail.cpp

```diff
@@ -211,24 +211,56 @@ static unsigned short getHeight(const float fx, const float fy, const float fz,
 	unsigned short h = hp.data[ix+iz*hp.width];
 	if (h == RC_UNSET_HEIGHT)
 	{
+		// Search in a spiral
+		const int rad = rcMax(1, (int)ceilf(maxEdgeError));
+
+		int x = 1, z = 0, dx = 1, dz = 0;
+		int maxSize = rad * 2 + 1;
+		int maxIter = maxSize * maxSize - 1;
+		int nextRingIterStart = 8;
+		int nextRingIters = 16;
+
 		// Special case when data might be bad.
 		// Find nearest neighbour pixel which has valid height.
-		const int off[8*2] = { -1,0, -1,-1, 0,-1, 1,-1, 1,0, 1,1, 0,1, -1,1};
 		float dmin = FLT_MAX;
-		for (int i = 0; i < 8; ++i)
+		for (int i = 0; i < maxIter; i++)
 		{
-			const int nx = ix+off[i*2+0];
-			const int nz = iz+off[i*2+1];
-			if (nx < 0 || nz < 0 || nx >= hp.width || nz >= hp.height) continue;
-			const unsigned short nh = hp.data[nx+nz*hp.width];
-			if (nh == RC_UNSET_HEIGHT) continue;
-			
-			const float d = fabsf(nh*ch - fy);
-			if (d < dmin)
+			const int nx = ix + x;
+			const int nz = iz + z;
+
+			if (nx >= 0 && nz >= 0 && nx < hp.width && nz < hp.height)
+			{
+				const unsigned short nh = hp.data[nx + nz*hp.width];
+				if (nh != RC_UNSET_HEIGHT)
+				{
+					const float d = fabsf(nh*ch - fy);
+					if (d < dmin)
+					{
+						h = nh;
+						dmin = d;
+					}
+				}
+			}
+
+			// Abort the search if we're entering a new 'ring' and have found a result
+			if (i + 1 == nextRingIterStart)
+			{
+				if (h != RC_UNSET_HEIGHT)
+					break;
+
+				nextRingIterStart += nextRingIters;
+				// Each successive 'ring' has 8 additional cells in it
```

Comment:
Basically the rings I'm talking about are these:
![grid](https://cloud.githubusercontent.com/assets/7887810/12331381/8d4b41d2-bae9-11e5-9e30-3d8a307292b4.png)
I don't know a better word for it. Maybe bounds?
The blue 'ring' has 8 cells. The red has 16. The green has 24, and so on. Each time we 'expand' the ring we add 2 cells around the center of each side, so we add 8 cells in total.
I added the check to minimize the impact of this change. Basically if we find a height at `distance = 1`, we stop there already, instead of continuing to outer rings. This means the change will abort here, for the vast majority of cases, and it will be the exact same behavior as before.
The only cases we need to continue the search are the rare ones, where we didn't find a height in the first ring.
Does that make sense? In any case I will add a better comment here.